Blame: options to git-diff to get line in previous commit

gerhardol opened the pull request:
## Proposed changes

GetOriginalLineInPreviousCommit() (to get line number in the previous commit)
must override the git diff option from git-config to follow the user preference
and to give predictable options for tests.
Similar, rename and whitespace options must be the same for git-diff
and git-blame.

* Include current section in checks
Better matching.

* BlameParserTest should not assume clone folder name

--

was

BlameParserTest should not assume clone folder

Follow up to #11434 Test fals if clonefolder is not gitextensions. I use various worktree folders with other names, so the test fails.
There is no reason to test the name of the clone folder.

Tests also fails locally due to matchingBlameLineNumber=2291, separate isue. (blame UI is OK, so are AppVeyor tests)

## Test methodology <!-- How did you ensure quality? -->

Removed incorrect test case

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

pmiossec reviewed (approved):
I have hesitated on that and should have thought about worktrees...

gerhardol commented:
@pmiossec The test fail locally still.
With f06e574547ae115fcf2356ddbcf9a300db6075ae
it works both locally and globally. 
Opinion to include?

mstv commented on IntegrationTests/UI.IntegrationTests/UserControls/GitBlameParserTest.cs:
```diff
@@ -14,7 +14,6 @@ public void GetOriginalLineInPreviousCommit()
         // line content is: internal bool ExecuteCommand(Command cmd)
 
         string gitExtensionsRepoPath = Path.GetFullPath(@"..\..\..\..\..\..\..");
-        Assert.IsTrue(gitExtensionsRepoPath.EndsWith("gitextensions"));
```
Check for existence of e.g. `GitExtensions.sln` **in** `gitExtensionsRepoPath`?

gerhardol replied:
That would replace this check, but I do not see that as adding any value.

mstv replied:
It would clearly indicate if the test is run from an unexpected folder. But there is no urgent need for this check. So, as you prefer.

RussKie replied:
Is there a reason we can't leverage `TestContext`?
![image](https://github.com/gitextensions/gitextensions/assets/4403806/2b1ea59a-3192-4eb1-9f8d-49baa586778d)

pmiossec commented:
> @pmiossec The test fail locally still. With [f06e574](https://github.com/gitextensions/gitextensions/commit/f06e574547ae115fcf2356ddbcf9a300db6075ae) it works both locally and globally. Opinion to include?

It seems legit but I really have no idea why it is failing for you locally as it should behave the exact same way for everyone (it is passing for me).
I'm very interested by seeing the output if the diff command...

gerhardol commented:
> > @pmiossec The test fail locally still. With [f06e574](https://github.com/gitextensions/gitextensions/commit/f06e574547ae115fcf2356ddbcf9a300db6075ae) it works both locally and globally. Opinion to include?
> 
> It seems legit but I really have no idea why it is failing for you locally as it should behave the exact same way for everyone (it is passing for me). I'm very interested by seeing the output if the diff command...

I do not see why the test fails either, but the change is what should be used what I see.

Also add one more assert?

```c#
        matchingBlameLineNumber = new GitBlameParser(() => gitModule)
                    .GetOriginalLineInPreviousCommit(selectedBlamedRevision, "GitUI/CommandsDialogs/FormBrowse.cs", 2090);
        Assert.AreEqual(2085, matchingBlameLineNumber); // line content is still arond: internal enum Command
```
 git -c color.ui=never -c diff.submodule=short -c diff.noprefix=false -c diff.mnemonicprefix=false -c diff.ignoreSubmodules=none -c core.safecrlf=false diff -U0 187faba05df2462bc817194740573795f2c92c3b 52476f30670ba5338756b606841fb0a346fd6460 -- GitUI/CommandsDialogs/FormBrowse.cs > t.txt

[t.txt](https://github.com/gitextensions/gitextensions/files/13797562/t.txt)

- Git Extensions 33.33.33
- Build df267e15036f0bb550f06f627d8d71eed0f85e77
- Git 2.43.0.windows.1
- Microsoft Windows NT 10.0.19045.0
- .NET 8.0.0
- DPI 96dpi (no scaling)
- Portable: False
- Microsoft.WindowsDesktop.App Versions

```
    Microsoft.WindowsDesktop.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 5.0.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 6.0.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 6.0.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 7.0.14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 8.0.0-rc.2.23479.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

pmiossec commented:
> I do not see why the test fails either,

I have tested with your output (which is the same than me) and it is working well. So I really don't get what is not working on you local env.

> but the change is what should be used what I see.

yes so all good for me. Not sure I would like to spend time investigating more...


> Also add one more assert?
> 
> ```cs
>         matchingBlameLineNumber = new GitBlameParser(() => gitModule)
>                     .GetOriginalLineInPreviousCommit(selectedBlamedRevision, "GitUI/CommandsDialogs/FormBrowse.cs", 2090);
>         Assert.AreEqual(2085, matchingBlameLineNumber); // line content is still arond: internal enum Command
> ```

Yes, good idea. It check the edge case you fixed.

pmiossec commented:
@gerhardol I have push a little refactoring.
Results are consistent with AppVeyor for me, the 2nd test is failing locally.
I have investigated a little (data, diff and git settings) and I don't have a clue why we don't have the same behavior.
But regarding the failing test, I need to investigate more to confirm but your value seems more accurate than 2149 that AppVeyor and me got.

pmiossec commented on IntegrationTests/UI.IntegrationTests/UserControls/GitBlameParserTest.cs:
```diff
@@ -1,11 +1,28 @@
+    [OneTimeSetUp]
+    public void OneTimeSetUp()
+    {
+        _useHistogramDiffAlgorithm = AppSettings.ProvideAutocompletion;
```
Is it a copy/paste error? Or could you explain what is the link between AppSettings.ProvideAutocompletion and AppSettings.UseHistogramDiffAlgorithm?

gerhardol replied:
typo, fixed...

pmiossec commented:
In the end, you fixed a totally different and more ambitious issue (i.e. the lack of settings support) than the one originally expected to fix.
I think that the commit message should be adapted accordingly as long as the PR title and description...

gerhardol commented:
Squashed, rebased and added a change for comments (and remove an option to git-blame that has no effect).
Also handled the other options, as well as changed the "multiline" test

Hopefully OK now...

mstv commented on GitCommands/Git/GitModule.cs:
```diff
@@ -3066,9 +3066,9 @@ public GitBlame Blame(string? fileName, string from, Encoding encoding, string?
-                { AppSettings.DetectCopyInFileOnBlame, "-M" },
-                { AppSettings.DetectCopyInAllOnBlame, "-C" },
-                { AppSettings.IgnoreWhitespaceOnBlame, "-w" },
+                { AppSettings.DetectCopyInFileOnBlame, "-M" }, // as git-diff --find-renames
+                { AppSettings.DetectCopyInAllOnBlame, "-C" }, // as git-diff --find-copies
+                { AppSettings.IgnoreWhitespaceOnBlame, "-w" }, // as git-diff --ignore-all-space
```
For those who wonder (and have not read the comments in the next file yet): "git blame" does not support the speaking / long arguments.

gerhardol replied:
AppSettings name is hopefully mostly describing what is done. 
One reason long options are in comments is to simplify searches

pmiossec replied:
> AppSettings name is hopefully mostly describing what is done.

Yes, naming is good enough for me and I was wondering why exactly the comments but the explanation given makes perfectly sense:

> One reason long options are in comments is to simplify searches

I just wonder if the settings property names could be (again) improved a little more:

* `DetectCopyInFileOnBlame` to `DetectCopyWithinFileOnBlame`
* `DetectCopyInAllOnBlame` to `DetectCopyInCommitOnBlame` or `DetectCopyInAllCommitFilesOnBlame`

Point of view?

gerhardol replied:
Not changed in this PR
